### DISCUSSION TITLE
Smart Meter Interface / Modbus TCP Issue

### DISCUSSION THREAD

**Discussion by torbenw84** [arendst/Tasmota#19042] [Support request] [answered]
Hi,
I am trying to query my SMA solar inverter using the new Modbus TCP feature in the smart meter interface but I was encountering some issues. 

1. The invester didn't respond to the requests. I started some analysis and found out that the message size in the request header sent from tasmota is wrong. The message size showed "01 00" instead of "00 06". I checked the coding and I think there is an error in the sml_swap() function. Currently it is defined like this:
 `uint16_t sml_swap(uint16_t in) {
  return (in << 8) || in >> 8;
}`

In my opinion the "logical or" is wrong and should be a "bitwise or":

`uint16_t sml_swap(uint16_t in) {
  return (in << 8) | in >> 8;
}`

I changed the code and now tasmota sends correct requests to the inverter and I finally receive responses!

The 2nd issue is, that I am not able to decode the responses. Always 0 is shown.
A respone message looks like this:
`03 04 04 00 00 01 44`

My meter config is:
+1,[192.168.178.80],m,0,502,SMA,1,50,03047837
1,030404UUuuUUuu@i0:1,PV power,W,pv1,0

Does anyone has an idea what is wrong?

Best Regards
Torben

**Comment by pkkrusty** [comment]
I am also receiving only zero even though dump shows good data coming in. Modbus TCP SML for Fronius inverter.

**Reply by torbenw84** [reply]
I have seen your issue and my decoding issue might be the same as yours. Maybe this is a general Modbus tcp issue.
Because of the possible bug in the first part of my comment, I decided to create a new discussion for this.

**Reply by gemu2015** [reply, marked as answer]
@torbenw84 
@pkkrusty 
yes the swap was a bug, i corrected this now and also fixed the receive issue. since i only have several serial modbus devices on no tcp modbus i could not test it properly.
please use my fork for further testing. if it works i will do a pr to tasmota

this test descriptor worked at least with a tcp modbus simulator 

```
>D
>B
=>sensor53 r
>M 1
+1,[192.168.188.39],m,0,502,SMA,0,10,02030000,02030001
1,020304UUuu@i0:10,Var 1,V,v1,1
1,020304UUuu@i1:10,Var 2,V,v2,1
#
```

**Reply by torbenw84** [reply]
@gemu2015 Thank you for the quick fix. I will do a test tomorrow!

**Comment by pkkrusty** [comment]
Quick test shows success. Thanks as usual for excellent support.
<img width="501" alt="Screen Shot 2023-07-06 at 07 53 15" src="https://github.com/arendst/Tasmota/assets/79770016/e1830e4f-a630-4450-a431-cb9b637ed9c1">

**Reply by pkkrusty** [reply]
Meter definition for Fronius Symo 10.0-3-M with important values listed:
```
>M 1
+1,[192.168.3.23],m,0,502,mrcyshp,1,10,r01039C870015
1,01032aUUuu@i0:100,AC Current,A,AC_Current,2
1,01032ax2UUuu@i0:100,Phase 1 Current,A,Phase_1_Current,2
1,01032ax4UUuu@i0:100,Phase 2 Current,A,Phase_2_Current,2
1,01032ax6UUuu@i0:100,Phase 3 Current,A,Phase_3_Current,2
1,01032ax16UUuu@i0:100,Phase 1 Voltage,A,Phase_1_Voltage,1
1,01032ax18UUuu@i0:100,Phase 2 Voltage,A,Phase_2_Voltage,1
1,01032ax20UUuu@i0:100,Phase 3 Voltage,A,Phase_3_Voltage,1
1,01032ax24UUuu@i0:10,Output Power,W,Output_Power,0
1,01032ax28UUuu@i0:100,Frequency,Hz,Frequency,2
1,01032ax36UUuu@i0:100,Temperature,C,Temperature,2
#
```

**Reply by gemu2015** [reply]
@torbenw84 

a friend of mine also has SMA, you may try his descriptor  (for U32,U64 etc support use my latest fork)

```
>M 1
+1,[192.168.56.91],m,0,502,SMA,0,10,03047741,03047747,03047777,03047831,03047833,03047835,030478ED,030478EF,030478F1,03047893,030478E9,0304787D,03047881,r03047AA50004,r03047AA90004
1,030404U32@i0:1000,Gesamtertrag,kWh,v1,3
1,030404U32@i1:1000,Tagesertrag,kWh,v2,3
1,030404U32@i2:1000,Einspeisung_ges,kWh,v3,3
1,030404S32@i3:100,DC Str. A,A,v4,2
1,030404S32@i4:100,DC Sp. A,V,v5,2
1,030404S32@i5:100,DC Le. A,W,v6,2
1,030404S32@i6:100,DC Str. B,A,v7,2
1,030404S32@i7:100,DC Sp. B,V,v8,2
1,030404S32@i8:100,DC Le. B,W,v9,2
;
1,030404S32@i9:100,AC Le.,W,v10,2
;
1,030404S32@i10:10,WR_Temp,°C,v11,2
;
1,030404U32@i11:1,Batterieladung,%%,v12,0
1,030404S32@i12:10,Batt_Temp,°C,v13,2
;
1,030408U64@i13:1000,Batt_Ladung,kWh,v14,3
1,030408U64@i14:1000,Batt_EntLadung,kWh,v15,3
#
```

**Reply by torbenw84** [reply]
@gemu2015 I just tested it and it finally also works for me. Thank you for the quick fixes and for the implementation of the modbus tcp protocol. I think this feature is very useful for a lot of people!

**Reply by 2000ede** [reply]
Hello @pkkrusty 
https://github.com/arendst/Tasmota/discussions/19042#discussioncomment-6371720

I wanted to use your code for a fronius inverter Fronius Symo 5.0-3-M. This also sends data to another system via TCP mode.

I took the code and just swapped the correct IP address. Unfortunately it doesn't show any data. Do I need to take anything else into the code?

I have already successfully read out another Modbus TCP with Tasmota

Thanks for help!

**Reply by pkkrusty** [reply]
Maybe your Modbus address is not set correctly in the Fronius? Here is the my current production code:

```
+1,[192.168.3.31],m,0,502,clinic,1,20,r01039C870015
1,=so3,256
1,01032aUUuu@i0:1,AC Current1,A,AC_Current1,0
1,01032ax2UUuu@i0:1,Phase 1 Current,A,Phase_1_Current,0
1,01032ax4UUuu@i0:1,Phase 2 Current,A,Phase_2_Current,0
1,01032ax6UUuu@i0:1,Phase 3 Current,A,Phase_3_Current,0
1,01032ax8SSss@i0:1,Curr Scale Fctr,SF,Curr_SF,0
1,01032ax16UUuu@i0:1,Phase 1 Voltage,A,Phase_1_Voltage,0
1,01032ax18UUuu@i0:1,Phase 2 Voltage,A,Phase_2_Voltage,0
1,01032ax20UUuu@i0:1,Phase 3 Voltage,A,Phase_3_Voltage,0
1,01032ax22SSss@i0:1,Vltg Scale Fctr,SF,Vltg_SF,0
1,01032ax24UUuu@i0:1,Output Power,W,Output_Power,0
1,01032ax26SSss@i0:1,Pwr Scale Fctr,SF,Pwr_SF,0
1,01032ax28UUuu@i0:1,Frequency,Hz,Frequency,0
1,01032ax30SSss@i0:1,Freq Scale Fctr,SF,Freq_SF,0
1,01032ax36UUuu@i0:1,Temperature,C,Temperature,0
1,01032ax38SSss@i0:1,Temp Scale Fctr,SF,Temp_SF,0
```
But also maybe you're causing conflict by having two TCP connections.  I would say try with only the Tasmota and see if you get data.

**Comment by Noschvie** [comment]
@2000ede do you have unique client ids configured? Maybe you can use qModMaster for tests.
